src/Pact/Eval.hs

```diff
@@ -436,6 +467,82 @@ resolveGovernance solvedDefs m' = fmap MDModule $ forM m' $ \g -> case g of
       Nothing -> evalError (_tInfo g) "Unknown module governance reference"
     _ -> evalError (_tInfo g) "Invalid module governance, should be var"
 
+-- Auxiliary data types for strict fold on heap cost.
+data HeapMemState
+  = HeapMemState
+  { _hmMemoEnv :: !(M.Map Name Bytes)
+  , _hmTotalMem :: !Bytes
+  }
+
+data HeapFold
+  = HeapFold
+  { _hfAllDefs :: !(HM.HashMap Text Ref)
+  , _hfMemoEnv :: !(M.Map Name Bytes)
+  , _hfTotalMem :: !Bytes
+  , _hfLastGas :: !Gas
+  }
+
+
+-- Inline the defuns according to the set heap limit for modules.
+-- We keep a memoized cost of each inlined `defun` so as to not calculate `sizeOf` more than once
+-- per inlined function.
+dresolveMem
+  :: Gas
+  -> Info
+  -> HeapFold
+  -> (Term (Either Text (Ref' (Term Name))), Text, c)
+  -> Eval e HeapFold
+dresolveMem gasLimit info (HeapFold allDefs costMemoEnv currMem _) (defTerm, defName, _) = do
+  let
+    (!unified, (HeapMemState costMemoEnv' totalMem))
+      = flip runState (HeapMemState costMemoEnv (sizeOf defTerm+currMem)) $ traverse (unifyMemo allDefs) defTerm
+  case unified of
+    t@TConst{} -> do
```

Comment:
Note (and maybe needs comment?): 

`TConst`s here are being evaluated in proper dependency order, so they do not need recursive reduction.

The change cannot be backported since it impacts gas.